lib/puppet/provider.rb

```diff
@@ -592,6 +592,14 @@ def <=>(other)
   # @return [void]
   # @api public
 
+  # @comment Document pre_resource_eval here as it does not exist anywhere else
+  # (called from transaction if implemented) @!method self.pre_resource_eval()
+  # @abstract A subclass may implement this - it is not implemented in the
+  # Provider class This method may be implemented by a provider in order to
+  # perform any setup actions needed, such as creating a handle. It will be
+  # called at the beginning of the transaction if the provider has implemented
+  # the method @return [void]
+
```

Comment:
About what I was mentioning, it'd be good to include this API doc change, the call to ` prov_class.pre_resource_eval if prov_class.respond_to?(:pre_resource_eval)` and a spec test that exercises the new method in a new PR. Then everything else can stay here for the selinux changes.

Reply:
How does this look? I think the spec tests will pass once this PR is merged: https://github.com/puppetlabs/puppet/pull/9376/files

Reply:
I think I need to move the `should call Selinux.selinux_close in case Selinux is enabled` test in transaction_spec.rb to the new PR too since that test fails without the `prov_class.pre_resource_eval if prov_class.respond_to?(:pre_resource_eval)` call

Reply:
I think I'd either need to move `should call Selinux.selinux_close in case Selinux is enabled` from transaction_spec.rb to the new PR or keep the `prov_class.pre_resource_eval if prov_class.respond_to?(:pre_resource_eval)` in this PR to get all spec tests to pass